### DISCUSSION TITLE
Problem when starting MotoROS2

### DISCUSSION THREAD

**Discussion by alexandraH99** [Yaskawa-Global/motoros2#200] [Troubleshooting]
Hi, I'm using YRC1000 controller and ROS 2 Humble.

I followed all the instructions with no problem showing up until [verifying successful installation](https://github.com/Yaskawa-Global/motoros2#verifying-successful-installation). When I started the micro_ros_agent, it got stuck at logger setup, with no error popping up. I tried both the docker image and ros2 package, the results were the same. 

I'm assuming the problem is with the connection between the controller and the pc, but if I ping the controller's ip address it works fine. I'm wondering how to troubleshoot this specific problem. Thanks!

<img width="1115" alt="Screenshot 2024-01-10 at 3 43 16 PM" src="https://github.com/Yaskawa-Global/motoros2/assets/107330754/ac0a521c-4833-4353-909a-ed100ae5f93b">

**Comment by gavanderhoorn** [comment]
> it got stuck at logger setup, with no error popping up

are you sure it's actually *stuck*?

At the normal debug level (ie: `info`) I'm pretty sure that's the only output you see, until MotoROS2 connects to the Agent.

If you've configured the correct Agent IP address and port in the configuration `.yaml` *and* uploaded that configuration to your YRC1000, you should see MotoROS2 connecting to the Agent running on your PC.

If you believe you did configure everything correctly, but are not seeing any connection attempts, please [start the debug log listener](https://github.com/Yaskawa-Global/motoros2/blob/9026a1d0b50b483f7c334bfd5ebbac8b328bfdb6/doc/troubleshooting.md#debug-log-client) and restart your controller. The log file should be created in the directory you start the debug listener in.

Attach that file here and we can check what MotoROS2 is trying to do.

**Reply by alexandraH99** [reply]
Thanks for the quick response. I tried the debug listener before but it said "no messages received", which made me think MotoRos2 was not configured properly. However, I wasn't able to find what the problem was. And the IP address and port were correct.

**Reply by gavanderhoorn** [reply]
> I tried the debug listener before but it said "no messages received"

just making sure: you did reboot the controller after starting the debug listener, correct? So turned it off and then on again.

> which made me think MotoRos2 was not configured properly. However, I wasn't able to find what the problem was. And the IP address and port were correct.

Can you please attach your configuration file in a comment here? It's the `motoros2_config.yaml` file you copied to the controller in [this step](https://github.com/Yaskawa-Global/motoros2/tree/9026a1d0b50b483f7c334bfd5ebbac8b328bfdb6#yrc1000-and-yrc1000-micro).

Please also attach a copy of the `PANELBOX.LOG` file (you can copy it to a USB stick using the pendant).

And just to check something else: could you show the output of `ufw status` if you run that in a terminal on your ROS PC?

**Reply by alexandraH99** [reply]
> just making sure: you did reboot the controller after starting the debug listener, correct? So turned it off and then on again.

I first started mirco_ros_agent when the robot was off, then ran the debug listener, and finally turned on the controller.

> could you show the output of ufw status if you run that in a terminal on your ROS PC?

It's `Status: inactive`

The files are attached here:
[PANELBOX.LOG](https://github.com/Yaskawa-Global/motoros2/files/13894117/PANELBOX.LOG)
[motoros2_config.yaml.zip](https://github.com/Yaskawa-Global/motoros2/files/13894121/motoros2_config.yaml.zip)

**Reply by gavanderhoorn** [reply]
From your `PANELBOX.LOG` it's unclear whether the MotoROS2 `.out` was correctly installed.

I would have expected a line like the following in your `PANELBOX.LOG`:

> ```
> //MOTOPLUS APP
> 01: MotoROS2 (humble)                0.1.2            micro-ROS based ROS 2 interface 
> ```

Could you confirm you followed the steps in [this section](https://github.com/Yaskawa-Global/motoros2/tree/9026a1d0b50b483f7c334bfd5ebbac8b328bfdb6#dx200-yrc1000-and-yrc1000micro), specifically:

> In *Maintenance* mode:
> 
>  1. upgrade to *MANAGEMENT* security level by touching `[System Info]`→`[Security]` (default password is all `9`'s)
>  1. touch `[MotoPlus APL]`→`[Device]` to select CF, SD, or USB memory type
>  1. touch `[MotoPlus APL]`→`[Load (User App)]` to select and load the `mr2_*_*.out` file

Could you (also in *Maintenance* mode) check the application is installed:

>  1. touch `[MotoPlus APL]`→`[File List]` and verify that MotoROS2 was properly installed and no other MotoPlus applications are currently loaded on the controller

?

If the `.out` doesn't show up on that list, MotoROS2 is not installed, which would explain why the debug listener doesn't receive any messages and why you're not seeing any connection attempts in the Agent's log.

**Reply by alexandraH99** [reply]
"mr2_yrc1_h.out" is on that list, and it is the only file on that list

**Reply by ted-miller** [reply]
Please check robot parameter `S2C1100` and make sure it has a value of 1. This would control whether the application attempts to startup at bootup time.

**Reply by alexandraH99** [reply]
It does have the value of 1

**Reply by ted-miller** [reply]
Also, in Normal mode, you can go to `MotoPlus APL` > `MotoPlus MONITOR`

![image](https://github.com/Yaskawa-Global/motoros2/assets/4623811/e6f886b8-11a1-49a4-ba55-b8b6fd99b05c)

Then go to `Application Run Flow` and make sure that mr2 is in the Active Application column.

![image](https://github.com/Yaskawa-Global/motoros2/assets/4623811/676c8752-9a6e-4ebf-97a5-fe2dfbf240eb)

**Reply by gavanderhoorn** [reply]
Hm.

Did you see the [8001[10]](https://github.com/Yaskawa-Global/motoros2/blob/9026a1d0b50b483f7c334bfd5ebbac8b328bfdb6/doc/troubleshooting.md#alarm-800110) alarm on the pendant when you first rebooted the controller after installing the `.out`?

That would tell us whether the application is at least getting started.

If you did get the alarm, but aren't seeing any alarms now, I'm not sure what's going on.

@ted-miller: any ideas? Could there be parameters preventing the application from starting?

**Reply by ted-miller** [reply]
> Did you see the [8001[10]](https://github.com/Yaskawa-Global/motoros2/blob/9026a1d0b50b483f7c334bfd5ebbac8b328bfdb6/doc/troubleshooting.md#alarm-800110) alarm on the pendant when you first rebooted the controller after installing the .out?

More specifically, were there _any_ alarms at startup?

**Reply by alexandraH99** [reply]
> Then go to Application Run Flow and make sure that mr2 is in the Active Application column.

mr2 is in the Active Application column.

> Did you see the [8001[10]](https://github.com/Yaskawa-Global/motoros2/blob/9026a1d0b50b483f7c334bfd5ebbac8b328bfdb6/doc/troubleshooting.md#alarm-800110) alarm on the pendant when you first rebooted the controller after installing the .out?

Actually this alarm never showed up.

> More specifically, were there any alarms at startup?

All the alarms I saw were illustrated in the tutorial. The only difference is I never saw 8001[10].

**Reply by gavanderhoorn** [reply]
> All the alarms I saw were illustrated in the tutorial.

do you remember which alarms you did see?

**Reply by alexandraH99** [reply]
I saw 8013[0], which was expected. When uploading the new config file, I saw [ERROR 4120](https://github.com/Yaskawa-Global/motoros2/blob/9026a1d0b50b483f7c334bfd5ebbac8b328bfdb6/doc/troubleshooting.md#error-4120-concurrent-io-memory-is-full), and I solved it following the troubleshooting tutorial.

**Reply by gavanderhoorn** [reply]
@ted-miller: without debug listener output this is hard to verify, but I'm wondering whether for some reason the `.yaml` got corrupted and now prevents MotoROS2 from starting up.

@alexandraH99: could you perhaps try to start with a fresh copy of `motoros2_config.yaml`, only edit the Agent IP and port then follow the [Updating the configuration](https://github.com/Yaskawa-Global/motoros2/tree/9026a1d0b50b483f7c334bfd5ebbac8b328bfdb6#updating-the-configuration) steps?

If you really want to start fresh, you could reinitialise the SRAM drive, similar to what you already did. to fix `4120` -- but then you'd have to [upload the file again](https://github.com/Yaskawa-Global/motoros2/tree/9026a1d0b50b483f7c334bfd5ebbac8b328bfdb6#yrc1000-and-yrc1000-micro) instead of updating it.

**Reply by ted-miller** [reply]
> without debug listener output this is hard to verify, but I'm wondering whether for some reason the .yaml got corrupted and now prevents MotoROS2 from starting up

There's other output before parsing the file. 
Plus the .out file should appear in the PANELBOX.

The M+ definitely isn't running.

@alexandraH99:
Could you save out a CMOS.BIN file from maintenance mode?
`Ext Memory` > `Save` > SELECT `CMOS.BIN` > ENTER
Also, if you're in North/South America, please let me know the warranty ID. It should be on a sticker on the controller cabinet. I can lookup how your controller was configured when it shipped.

**Reply by alexandraH99** [reply]
@gavanderhoorn  I started fresh and uploaded the new yaml file, but I still didn't get alarm 8001[10]

**Reply by alexandraH99** [reply]
@ted-miller The warranty ID is 783725-2-1
Attached is the file:
[CMOS.BIN.zip](https://github.com/Yaskawa-Global/motoros2/files/13895098/CMOS.BIN.zip)

**Reply by ted-miller** [reply]
@alexandraH99 
I looked through the alarm history in your CMOS. The problem is caused by the alarm `1020 [6] APPLICATION LOAD ERROR`. The .out file is failing to load at startup.

Possible causes:
1. The .out file may have been corrupted. Please delete the .out file from your computer, usb drive, and the teach pendant (maintenance mode). Then download a new copy and load it into the robot controller.
2. If you have a USB drive connected to the controller's CPU board (CN102), please make sure that there aren't any .out files on that drive.
3. In Maintenance mode, go to `File` > `Initialize` > `System Data` > `MotoPlus Temporary File  MPAPP.TMP`. Select this and press Enter to initialize it.

**Reply by gavanderhoorn** [reply]
I'd suggest to [verify integrity of the downloaded file](https://github.com/Yaskawa-Global/motoros2/tree/9026a1d0b50b483f7c334bfd5ebbac8b328bfdb6#verifying-integrity-of-the-download). That should tell you whether the `.out` is OK or not.

**Reply by alexandraH99** [reply]
@ted-miller I've started fresh, downloading everything again into a USB stick with nothing else in it, installing .out file following the tutorial and Step 3, and uploading the config file with IP address and port changed only. However, I still can't get alarm 8001[10].

@gavanderhoorn And yes I've verified the integrity of .out file.

**Reply by gavanderhoorn** [reply]
Do you see the error @ted-miller mentioned (ie: the `1020[6]`)? Or any other errors/warnings?

**Reply by alexandraH99** [reply]
@gavanderhoorn I've never seen that alarm popped up on the pendant. Is there a way to check it without going through CMOS?

**Reply by gavanderhoorn** [reply]
On the pendant: `SYSTEM INFO` > `ALARM HISTORY`.

**Reply by alexandraH99** [reply]
@gavanderhoorn No error from today shows up in `SYSTEM INFO` > `ALARM HISTORY`

**Reply by gavanderhoorn** [reply]
Any other errors/warnings which seem interesting?

It does look like the `.out` is just not getting run.

@ted-miller: anything you can think of we could remotely diagnose?

I've checked again and `PANELBOX.LOG` shows `MotoPlus FUNC.: USED` and `MOTOMAN DRIVER: USED`, so that should be setup.

**Reply by alexandraH99** [reply]
@gavanderhoorn All the alarms after I started trying to install MotoRos 2 listed in `SYSTEM INFO` > `ALARM HISTORY` are `1020[6]` followed by `1050[1]` every time. None of them are from today, after I started fresh. Other than those, all the alarms I've seen on the pendant are listed in the tutorial and have corresponding solutions. (I'm not sure why these alarms don't show up in the history, maybe because they are resolved?)

**Reply by gavanderhoorn** [reply]
Ah, you may only be looking at the `MAJOR` alarms. Use the `PAGE` button at the bottom of the screen to change to the other categories.

**Reply by alexandraH99** [reply]
I've checked the other pages. The only error after I started fresh today is `8013[0]`, which is expected

**Reply by gavanderhoorn** [reply]
@alexandraH99 just checking: have you been able to make any progress?

@ted-miller: would you have any ideas for next steps?

**Reply by ted-miller** [reply]
`8013 [0]` would indicate that the MotoROS2 application is running and operating as expected. So I'm assuming that the `1020 [6] APPLICATION LOAD ERROR` has been resolved.

**Reply by alexandraH99** [reply]
@gavanderhoorn Not really. We are now using other methods to communicate to the robot, such as MotoRos 1 and customized MotoPlus applications. (We have no trouble switching between those applications.)

**Reply by gavanderhoorn** [reply]
Would you happen to have VGA connector kit for your YRC1000?
